Binary domain: boundary conditions and cleanup

nilsdeppe opened the pull request:
## Proposed changes

- Quite a bit of cleanup and simplifications in the tests. This was largely motivated by me wanting to make adding the boundary conditions easier for myself
- Remove the equiangular option since the resulting domain is incorrect
- Add boundary conditions to the binary domain
- Some clang-tidy stuff, yay

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Scope variables in CompactObj domain creator

nilsdeppe committed:
Format BinaryCompactObject creator tests

nilsdeppe committed:
Remove equiangular option from BinaryCompactObject domain creator

The implementation is not fully functional yet

nilsdeppe committed:
Simplify BinaryCompactObject test

nilsdeppe committed:
Add boundary condition support to BinaryCompactObject domain creator

nilsdeppe committed:
Clang-tidy fixes in BinaryCompactObject domain creator

nilsvu commented on tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:
```diff
@@ -134,9 +134,10 @@ void test_connectivity() {
   }
 }
-}
+
```
This commit is just formatting, did you intend to make more changes in this commit? Else please reword the commit message.

nilsvu commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -264,7 +266,23 @@ class BinaryCompactObject : public DomainCreator<3> {
+
+  template <typename BoundaryConditionsBase>
+  struct OuterBoundaryCondition {
+    static std::string name() noexcept { return "OuterBoundaryCondition"; }
```
I think you meant to name this "InnerBoundary", and the one below "OuterBoundary", and put them in a `BoundaryConditionsGroup`, like you did for `Interval` (I like the grouping you did there).

fmahebert replied:
For consistency with Interval et al, return "OuterBoundary" here

nilsdeppe committed:
Use CHECK_THROWS to test PARSE_ERRORS in binary domain

Generally, having separate tests for checking error messages is very slow,
checking that the constructor throws reduces test time and is also easier to
write and maintain.

nilsdeppe commented:
I made the requested changes and pushed a new commit that replaces the error tests with `CHECK_THROWS` tests. Reducing the number of test cases will make running the unit tests faster, which, while not the bottleneck in CI, means people are more likely to run the tests locally.

fmahebert commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -146,6 +153,21 @@ BinaryCompactObject::BinaryCompactObject(
+       outer_boundary_condition_ != nullptr)) {
+    PARSE_ERROR(context,
+                "Must specify either both inner and outer boundary conditions "
+                "or neither.");
```
If I don't excise A or B (say for a BNS run) then what inner boundary condition do I want to pass? None? Or should we not require the inner BC in this scenario?

nilsdeppe replied:
I don't know how to not require a BC in this case because we would effectively need a "default" value. The only ways to really handle this that I can think of right now are:
1. have an `Alternatives<tmpl::list<ExciseA, ExciseB, InnerBoundaryCondition>, tmpl::list<>>`, but I don't think this will work because this is effectively setting a "default" value
2. leave as-is with a comment
3. decide to add a None boundary condition, either system-specific or general. Probably a general one like the periodic would work

kidder replied:
My suggestion is to:  
- Add a `None` BC that errors if it is ever used
- Either require that `None` is used if ExciseInteriorA and ExciseInteriorB are both false; or add a comment to InnerBoundaryCondition's help string saying that this option is ignored in that case

fmahebert replied:
I would be satisfied with a comment (as per @kidder's suggestion) for now, and perhaps adding a `None` BC in a separate PR (or making an issue for it).

fmahebert commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -264,7 +266,23 @@ class BinaryCompactObject : public DomainCreator<3> {
-  using options = tmpl::list<
+  template <typename BoundaryConditionsBase>
+  struct InnerBoundaryCondition {
+    static std::string name() noexcept { return "InnerBoundaryCondition"; }
```
For consistency with Interval et al, return "InnerBoundary" here ... or actually maybe "ExcisionBoundaries" is even better, to make clear it'll apply to any/all excision regions if present?

fmahebert commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -264,7 +266,23 @@ class BinaryCompactObject : public DomainCreator<3> {
+  struct InnerBoundaryCondition {
+    static std::string name() noexcept { return "InnerBoundaryCondition"; }
+    static constexpr Options::String help =
+        "Options for the inner boundary conditions.";
```
To make clearer this applied to all excision regions, I suggest one of,
- append a sentence like "This boundary condition will be applied to all inner excision boundaries", OR
- "inner" -> "excision region"

nilsdeppe commented:
Dependent on #2884

nilsdeppe commented:
I have squashed the previous fixup and pushed a new one that use `None` to do error checking when nothing is excised.

I did give `Options::Auto<...>` a try for not having this semi-redundant options, but this is non-trivial because `std::optional<std::unique_ptr>` behaves differently than `std::unique_ptr` in terms of implicit conversions to base class pointers. If someone wants to give that a try, I have no objections to using that design, but I can't really spend time figuring that out right now.

kidder commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -153,6 +154,15 @@ BinaryCompactObject::BinaryCompactObject(
+        not is_none(inner_boundary_condition_)))) {
+    PARSE_ERROR(
+        context,
+        "Inner boundary condition must be None if not excising both A and B");
```
I would say "Inner boundary condition must be None if ExciseInteriorA and ExciseInteriorB are both false" to directly cite the input file options.  If you do so, don't forget to update the corresponding text in the `Catch::Matchers::Contains`

fmahebert replied:
I like this suggestion 👍 

I would also suggest moving this `if` block down below the `if` blocks on lines 166–180. This has the aesthetic appeal of checking for simple mistakes before checking for complicated ones... and the code appeal that below line 166–173 you no longer need the innermost "inner == nullptr" check because you know "(outer == nullptr) == (inner == nullptr)" and you're in an `if` block where "outer != nullptr", which implies "inner != nullptr".

nilsdeppe replied:
While I mostly agree, then I need to rework a lot of logic in the test cases for what different iteration of the for loop triggers what error, and that turns out to be quite tedious and time consuming

kidder reviewed (commented):
just one small thing that can be squashed when @fmahebert and @nilsleiffischer also give the go ahead to squash

fmahebert commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -153,6 +154,15 @@ BinaryCompactObject::BinaryCompactObject(
+        context,
+        "Inner boundary condition must be None if not excising both A and B");
+  }
   if ((inner_boundary_condition_ != nullptr and
```
Option to write this as "(outer == nullptr) != (inner == nullptr)" instead... totally up to you I'm happy either way

nilsdeppe replied:
I find that much harder to understand, honestly :)

nilsdeppe replied:
Turns out there's a thing for this: `xor`. I've pushed a separate fixup so it's easily dropped if people don't like it

fmahebert replied:
Oh yeah 😅

fmahebert reviewed (commented):
You have my permission to squash, including these changes if any.

fmahebert commented:
LGTM

kidder commented:
@nilsleiffischer please look at the fixups

nilsvu reviewed (commented):
LGTM 👍 Please don't wait for my re-approval after this gets squashed.

nilsdeppe commented:
Rebased on develop and squashed in the fixups. I also squashed in `inner_bc_ == nullptr xor outer_bc_ == nullptr` -> `(inner_bc_ == nullptr) xor (outer_bc_ == nullptr)` because of a GCC warning

kidder commented:
@nilsleiffischer this has been squashed
